src/com/amazon/ionelement/api/Ion.kt

```diff
@@ -36,148 +36,384 @@ import com.amazon.ionelement.impl.SymbolElementImpl
 import com.amazon.ionelement.impl.TimestampElementImpl
 import java.math.BigInteger
 
-/** Returns a memoized instance of [IonElement] that represents an Ion `null.null` or a typed `null`. */
-fun ionNull(
-    elementType: ElementType = ElementType.NULL
-): IonElement =
-    ALL_NULLS.getValue(elementType)
+typealias Annotations = List<String>
+
+/*
+
+# Notes
+
+## vararg and collection constructors
+
+The vararg element collection constructors Kotlin API are not fully idiomatic from the perspective of Java, because
+Java requires variadic parameters to be the last parameter, but kotlin does not.  When a vararg parameter is defined
+in Kotlin that is not in the last one, it appears to Java to be simply an array.  Unfortunately, `ionListOf`,
+`ionSexpOf` and `ionStructOf` constructors all accept a vararg parameter as the first argument and this cannot change
+without breaking many clients.
+
+This each of these functions has an overload that from Java appears to accept either an IonElement[] or an
+Iterable<IonElement>.
+
+## @JvmOverloads
+
+@JvmOverloads is a very useful tool for Java compatibility that synthesizes overloads which specify the default values
+of your parameters when left unspecified by calling code, however, this do not generate overloads for all possible
+combinations of values.
+
+For example, the following definition:
+
+```
+@JvmOverloads
+fun ionInt(l: Long, annotations: Annotations = emptyList(), metas: MetaContainer = emptyMetaContainer()): IntElement = ...
+```
+
+Synthesizes the following overloads:
+
+```
+// Java syntax
+IntElement ionInt(Long l) { ... }
+IntElement ionInt(Long l, Annotations annotations) { ... }
+IntElement ionInt(Long l, Annotations annotations, MetaContainer metas) { ... }
+```
+
+Missing from this is an overload that accepts only the the `l` and `metas` parameters, which has to be added
+manually:
+
+```
+IntElement ionInt(Long l, MetaContainer metas) { ... }`
+```
+
+Below, we use a combination of @JvmOverloads and the manually implemented overloads for each Ion data type.
+*/
 
 /**
  * Creates an [IonElement] that represents an Ion `null.null` or a typed `null` with the specified metas
  * and annotations. */
+@JvmOverloads
```

Comment:
Did you want to also do the annotation that makes the class name not `IonKt`?

Reply:
`@file: JvmName("Ion")` already exists at the top of the file (not shown in the diff).